src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PeerRefresher.cs

```diff
@@ -17,61 +17,202 @@
+            }
+            else if (headAndParentHeaders.Length == 2)
+            {
+                // Maybe the head is not the same as we expected. In that case, leave it as null
+                if (headBlockhash == headAndParentHeaders[1].Hash)
+                {
+                    headBlockHeader = headAndParentHeaders[1];
+                }
+                headParentBlockHeader = headAndParentHeaders[0];
+            }
+            else if (headAndParentHeaders.Length > 0)
+            {
+                if (_logger.IsTrace) _logger.Trace($"PeerRefreshForFCU unexpected response length when fetching header: {headAndParentHeaders.Length}");
+            }
+
+            finalizedBlockHeader = await getFinalizedHeaderTask;
+        }
+        catch (TaskCanceledException exception)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace(
+                    $"PeerRefreshForFCU canceled for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+            _syncPeerPool.ReportRefreshCancelled(syncPeer);
+            throw;
+        }
+        catch (Exception exception)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine}{exception}");
+            _syncPeerPool.ReportRefreshFailed(syncPeer, "faulted");
+            // TODO: how to know if exception is transient
+            syncPeer.Disconnect(DisconnectReason.DisconnectRequested, "refresh peer info fault - faulted");
+            return;
+        }
+
+        if (_logger.IsTrace)
+        {
+            _logger.Trace($"PeerRefreshForFCU received block info from {syncPeer.Node:c}");
+            _logger.Trace($"PeerRefreshForFCU headHeader: {headBlockHeader}");
+            _logger.Trace($"PeerRefreshForFCU headParentHeader: {headParentBlockHeader}");
+            _logger.Trace($"PeerRefreshForFCU finalizedBlockHeader: {finalizedBlockHeader}");
+        }
+
+        if (finalizedBlockhash != Keccak.Zero && finalizedBlockHeader == null)
+        {
+            if (_logger.IsTrace)
+                _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine} - Finalized block header not found");
+            _syncPeerPool.ReportRefreshFailed(syncPeer, "no finalized block header");
+            return;
+        }
+
+        List<BlockHeader> headersToCheck = new [] {
+            headBlockHeader,
+            headParentBlockHeader,
+            finalizedBlockHeader
+        }.Where((header) => header != null).ToList();
+        
+        foreach (BlockHeader header in headersToCheck)
+        {
+            if (!HeaderValidator.ValidateHash(header))
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace($"PeerRefreshForFCU failed for node: {syncPeer.Node:c}{Environment.NewLine} Invalid block hash. Header: {header}");
+                _syncPeerPool.ReportRefreshFailed(syncPeer, "invalid header hash");
```

Comment:
With additional context in `reason` I think we don't need this log, it will duplicate.